webapp/channels/src/components/markdown/index.ts

```diff
@@ -34,7 +34,13 @@ function makeGetChannelNamesMap() {
         (state: GlobalState, props: Props) => props && props.channelNamesMap,
         (channelNamesMap, channelMentions) => {
             if (channelMentions) {
-                return Object.assign({}, channelMentions, channelNamesMap);
+                const mention = Object.keys(channelMentions)[0];
+                const displayName = (channelMentions[mention] as { display_name: string; }).display_name;
+                if(displayName!==channelNamesMap[mention]) {
+                    return channelNamesMap;
```

Comment:
You're actually going to want to combine `channelNamesMap` and `channelMentions` instead of just returning one or the other. 

Unfortunately, I'm having a hard time explaining how to combine them since it's a bit weird, so I had to write out how I think this should look. Hopefully the comments help explain what I'm trying to do here.
```ts
function makeGetChannelNamesMap() {
    return createSelector(
        'makeGetChannelNamesMap',
        getChannelNameToDisplayNameMap,
        (state: GlobalState, props: Props) => props && props.channelNamesMap,
        getCurrentTeam,
        (channelNamesMap, channelMentions, currentTeam) => {
            if (!channelMentions) {
                return channelNamesMap;
            }

            // Start with the display name of each channel on the current team that's currently loaded by the web app
            const result: ChannelNamesMap = {...channelNamesMap};

            for (const [name, entry] of Object.entries(channelMentions)) {
                if (!result[name]) {
                    // This channel hasn't been loaded by the web app, so use the display name from post.props
                    result[name] = entry;
                    continue;
                }

                // post.props.channel_mentions for newer posts contains the display name of the channel and the team
                // that it's on. If the link is to a channel on a different team, use the display name from post.props
                if (
                    typeof entry === 'object' &&
                    entry.team_name &&
                    currentTeam.name !== entry.team_name
                ) {
                    result[name] = entry;
                }
            }

            return result;
        },
    );
}
```

Let me know if you have any questions or if there's anything I could help explain better

Reply:
ChannelMentions variable always contains only one key value pair right?
Instead of using for loop we can directly use first key right?

Reply:
It can actually contain multiple pairs if there are multiple links in the post. Here's a post I made on my local instance with multiple links in it
```js
{
	"id": "3c96xiefwpbqpdzczhocz35sbh",
	"user_id": "emn1z31webdnujs91csa56pjre",
	"channel_id": "pnhqscaah3dtzkfrshpq4dq8py",
	"message": "This is a link to ~town-square and a link to ~off-topic",
	"props": {
		"channel_mentions": {
			"off-topic": {
				"display_name": "This Is Off Topic",
				"team_name": "yourteamhere"
			},
			"town-square": {
				"display_name": "Town Square",
				"team_name": "yourteamhere"
			}
		}
	}
}
```

If you were looking at the type definition and thought there was only one field because of that, it's using an [index signature](https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures) to define the type of a value in the `ChannelNamesMap` even though we don't know how many keys-value pairs are in it

Reply:
Got it will make changes

Reply:
@hmhealey can you please review now?